### PR TITLE
Add benchmark tests for storing Cardano transactions

### LINKED ISSUE #1656: Use SQLite transactions when inserting `Cardano Transactions` and `Block Range Roots`
## Why
We can have a better insertion throughput if we use transactions in SQLite (see https://stackoverflow.com/questions/1711631/improve-insert-per-second-performance-of-sqlite).
This can be used to accelerate the insertion on Cardano transactions and block range roots on the signer and aggregator.

## What
Create insert queries inside a transaction when importing the Cardano Transactions ~~and block range roots~~.

## How

- [x] Create a benchmark for current implementation
- [x] Add SQLite transactions when importing Cardano Transactions ~~and block range roots~~
- [x] Measure the performance optimization between the two versions

## After
- Assess the relevance of adding SQLite transactions when importing block range roots
- How can we deploy this optimizations on the other store of signer/aggregator (do we need to pool connections to database?)

### PR DESCRIPTION
**Pull request by dlachaume**
## Content
This PR includes benchmark tests for evaluating the performance of storing Cardano transactions. The PR also includes modifications that improve them.
The modifications consist of:
- Adding SQL transactions
- Rework of `expand` implementation of `WhereCondition` in sqlite module

## Pre-submit checklist

- Branch
  - [ ] Tests are provided (if possible)
  - [ ] Crates versions are updated (if relevant)
  - [ ] CHANGELOG file is updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## `store_transactions` with 10K transactions

- ### **_Current state_**
Average time: 154.12 ms
image

- ### **_With modifications: adding SQL transactions_**
Average time: 116.35 ms
image

- ### **_With modifications: rework of `expand` impl_**
Average time: 98.078 ms
image

- ### **_With modifications: adding SQL transactions and rework of `expand` impl_**
Average time: 60.628 ms
image

-----

## `store_transactions` with 100K transactions

- ### **_Current state_**
Average time: 1.5205 s
image

- ### **_With modifications: adding SQL transactions_**
Average time: 1.1405 s
image

- ### **_With modifications: rework of `expand` impl_**
Average time: 939.73 ms
image

- ### **_With modifications: adding SQL transactions and rework of `expand` impl_**
Average time: 568.32 ms
image

-----

## `store_transactions` with 250K transactions

- ### **_Current state_**
Average time: 3.6821 s
image

- ### **_With modifications: adding SQL transactions_**
Average time: 2.8220 s
image

- ### **_With modifications: rework of `expand` impl_**
Average time: 2.3101 s
image

- ### **_With modifications: adding SQL transactions and rework of `expand` impl_**
Average time: 1.4331 s
image

-----

## `store_transactions` with 500K transactions

- ### **_Current state_**
Average time: 7.3614 s
image

- ### **_With modifications: adding SQL transactions_**
Average time: 5.5974 s
image

- ### **_With modifications: rework of `expand` impl_**
Average time: 4.5745 s
image

- ### **_With modifications: adding SQL transactions and rework of `expand` impl_**
Average time: 2.8139 s
image

-----

## `store_transactions` with 1M transactions

- ### **_Current state_**
Average time: 14.772 s
image

- ### **_With modifications: adding SQL transactions_**
Average time: 11.004 s
image

- ### **_With modifications: rework of `expand` impl_**
Average time: 9.2098 s
image

- ### **_With modifications: adding SQL transactions and rework of `expand` impl_**
Average time: 5.5863 s
image

-----

## `store_transactions` with 10M transactions
- ### **_With modifications: adding SQL transactions and rework of `expand` impl_**
Average time: 55.304 s
image

## Issue(s)
Closes #1656

### CODE REVIEW

**Review by Alenar — APPROVED** [review]

File: mithril-aggregator/Cargo.toml [hunk 1]
```diff
@@ -9,6 +9,10 @@ homepage = { workspace = true }
 license = { workspace = true }
 repository = { workspace = true }
 
+[[bench]]
+name = "cardano_transactions"
```
Comment (jpraynaud) [inline mithril-aggregator/Cargo.toml:13]: Maybe this should be named `cardano_transactions_import`?

File: mithril-aggregator/benches/cardano_transactions.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        .collect()
+}
+
+fn bench_store_transactions(c: &mut Criterion) {
+    const NB_CARDANO_TRANSACTIONS: usize = 100000;
+    let runtime = tokio::runtime::Runtime::new().unwrap();
+
+    let mut group = c.benchmark_group("Store transactions");
+    group.bench_function("store_transactions", |bencher| {
+        bencher.to_async(&runtime).iter(|| async {
```
Comment (jpraynaud) [inline mithril-aggregator/benches/cardano_transactions.rs:47]: The benchmark would probably be more accurate with the transactions generation done separately

File: mithril-aggregator/benches/cardano_transactions.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                i as u64 + 1,
+                format!("block_hash-{}", i),
+                i as u64 + 2,
+            )
+        })
+        .collect()
+}
+
+fn bench_store_transactions(c: &mut Criterion) {
+    const NB_CARDANO_TRANSACTIONS: usize = 100000;
```
Comment (jpraynaud) [inline mithril-aggregator/benches/cardano_transactions.rs:42]: Maybe multiple values could be benchmarked up to at least 10/100 million records?

Also the value could be easier to read like this:
```suggestion
    const NB_CARDANO_TRANSACTIONS: usize = 100_000;
```

File: mithril-aggregator/src/database/repository/cardano_transaction_repository.rs [hunk 4]
```diff
@@ -242,12 +242,14 @@ impl TransactionStore for CardanoTransactionRepository {
     }
 
     async fn store_transactions(&self, transactions: Vec<CardanoTransaction>) -> StdResult<()> {
+        self.connection.execute("BEGIN TRANSACTION;")?;
```
Comment (jpraynaud) [inline mithril-aggregator/src/database/repository/cardano_transaction_repository.rs:245]: Is it a concern to execute a transaction on very high volume of transactions? Don't we need to do multiple transactions instead?

File: mithril-aggregator/benches/cardano_transactions.rs (hunk 3 above)
Comment (dlachaume) [reply mithril-aggregator/benches/cardano_transactions.rs]: Done from 10K -> 1M transactions, see results below.
Replaced as suggested:
`const NB_CARDANO_TRANSACTIONS: usize = 1_000_000;`

**Review by jpraynaud — APPROVED** [review]
LGTM :+1:

**Pull request merged**